File: app/services/twitter.py
from __future__ import annotations

import json
import logging
import re
from abc import ABC
from collections.abc import AsyncGenerator
from json import JSONDecodeError
from typing import Annotated, Any, ClassVar, TypeVar

import httpx
from fake_useragent import UserAgent
from httpx import codes
from jsonpath_ng.ext import parse
from pydantic import AfterValidator, ValidationError, validate_call

from app.services.models.base import BaseModel
from app.services.models.post import TwitterPost
from app.services.models.user import TwitterUser
from app.utils import find_key_recursive

logger = logging.getLogger(__name__)

# Search values here - https://x.com
DEFAULT_KWARGS = {
    'articles_preview_enabled': False,
    'c9s_tweet_anatomy_moderator_badge_enabled': True,
    'communities_web_enable_tweet_community_results_fetch': True,
    'creator_subscriptions_quote_tweet_preview_enabled': False,
    'creator_subscriptions_tweet_preview_api_enabled': True,
    'freedom_of_speech_not_reach_fetch_enabled': True,
    'graphql_is_translatable_rweb_tweet_is_translatable_enabled': True,
    'longform_notetweets_consumption_enabled': True,
    'longform_notetweets_inline_media_enabled': True,
    'longform_notetweets_rich_text_read_enabled': True,
    'responsive_web_edit_tweet_api_enabled': True,
    'responsive_web_enhance_cards_enabled': False,
    'responsive_web_graphql_exclude_directive_enabled': True,
    'responsive_web_graphql_skip_user_profile_image_extensions_enabled': False,
    'responsive_web_graphql_timeline_navigation_enabled': True,
    'responsive_web_media_download_video_enabled': False,
    'responsive_web_twitter_article_tweet_consumption_enabled': True,
    'rweb_tipjar_consumption_enabled': True,
    'rweb_video_timestamps_enabled': True,
    'standardized_nudges_misinfo': True,
    'tweet_awards_web_tipping_enabled': False,
    'tweet_with_visibility_results_prefer_gql_limited_actions_policy_enabled': True,
    'tweet_with_visibility_results_prefer_gql_media_interstitial_enabled': False,
    'tweetypie_unmention_optimization_enabled': True,
    'verified_phone_label_enabled': False,
    'view_counts_everywhere_api_enabled': True,
    'responsive_web_grok_analyze_button_fetch_trends_enabled': False,
    'premium_content_api_read_enabled': False,
    'profile_label_improvements_pcf_label_in_post_enabled': False,
    'responsive_web_grok_share_attachment_enabled': False,
    'responsive_web_grok_analyze_post_followups_enabled': False,
    'responsive_web_grok_image_annotation_enabled': False,
    'responsive_web_grok_analysis_button_from_backend': False,
    'responsive_web_jetfuel_frame': False,
    'rweb_video_screen_enabled': True,
    'responsive_web_grok_show_grok_translated_post': True,
}


type UserID = int | str

ENTRIES_ONLY = parse('$..entries[?(@.entryId =~ "^(?!(cursor-|messageprompt-))")]')
CURSOR_BOTTOM = parse('$..entries[?(@.entryId =~ "^(cursor-bottom)")]')


def encode_params(data: dict[str, Any]) -> dict[str, str]:
    result = {}

    for key, value in data.items():
        if isinstance(value, dict):
            encoded_value = json.dumps(
                {a: b for a, b in value.items() if b is not None},
                separators=(',', ':'),
            )
            result[key] = str(encoded_value)

    return result


class TwitterAPIError(IOError):
    def __init__(self, status_code: int, reason: str) -> None:
        super().__init__(status_code, reason)

        self.status_code = status_code
        self.reason = reason

    @classmethod
    def from_response(cls, response: httpx.Response) -> TwitterAPIError:
        return cls(response.status_code, response.text)

    def __str__(self) -> str:
        return f'Unexpected error from Twitter API: {self.status_code} {self.reason}'


class TwitterAPI:
    def __init__(
        self,
        cookies: dict[str, str],
        base_url: str = 'https://x.com/i/api/graphql/',
        user_agent: str | None = None,
    ):
        self.client = httpx.AsyncClient(
            base_url=base_url,
            cookies=cookies,
            headers={
                'User-Agent': user_agent or UserAgent().chrome,
                'x-twitter-client-language': 'en',
                'x-twitter-active-user': 'yes',
                'Authorization': 'Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA',
                'x-csrf-token': cookies['ct0'],
                'x-twitter-auth-type': 'OAuth2Session',
            },
        )

    @staticmethod
    def _prepare_params(
        data: dict[str, Any] | None = None,
        features: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        features_payload = {**DEFAULT_KWARGS}

        if features:
            features_payload.update(features)

        return {'variables': data or {}, 'features': features_payload}

    async def list_graphql_request(
        self,
        url: str,
        data: dict[str, Any] | None = None,
        features: dict[str, Any] | None = None,
    ) -> AsyncGenerator[list[dict[str, Any]]]:
        params = self._prepare_params(data, features)

        cursor = None

        if cursor is not None:
            params['variables']['cursor'] = cursor

        response = await self.client.get(url, params=encode_params(params))

        if not response.is_success:
            logger.error('Twitter API error: %s', response.text)
            raise TwitterAPIError.from_response(response)

        try:
            json_response = response.json()
        except JSONDecodeError as ex:
            raise TwitterAPIError.from_response(response) from ex

        logger.debug('JSON response: %s', json.dumps(json_response))

        entries = find_key_recursive(json_response, 'entries')

        filtered_entries = [
            entry
            for entry in entries
            if entry.get('entryId') is not None
            and not entry.get('entryId').startswith(('cursor-', 'messageprompt-'))
        ]

        logger.debug('Filtered entries: %s', filtered_entries)

        if filtered_entries:
            yield filtered_entries

        # If no entries - we at the end of query, no need to continue
        else:
            return

        cursor_entry = find_key_recursive(json_response, 'entryId', 'cursor-bottom')

        # If no cursor - we at the end of query
        if not cursor_entry:
            return

        cursor = cursor_entry['content']['itemContent']['value']

    async def get_graphql_request(
        self,
        url: str,
        data: dict[str, Any] | None = None,
        features: dict[str, Any] | None = None,
    ) -> dict[str, Any] | None:
        params = self._prepare_params(data, features)

        response = await self.client.get(url, params=encode_params(params))

        if response.status_code == codes.NOT_FOUND:
            return None

        if not response.is_success:
            raise TwitterAPIError.from_response(response)

        try:
            json_response = response.json()
        except JSONDecodeError as ex:
            raise TwitterAPIError.from_response(response) from ex

        logger.debug('JSON response: %s', json.dumps(json_response))

        return json_response

    @property
    def users(self) -> Users:
        return Users(self)

    @property
    def likes(self) -> Likes:
        return Likes(self)

    @property
    def posts(self) -> Posts:
        return Posts(self)


def extract_post_id(x: str) -> str:
    match = re.match(r'^https://x\.com/(.*)/status/(\d+)(\?s=.*)?$', x)

    if match is None:
        raise ValueError('Invalid post ID')

    return match.group(2)


T = TypeVar('T', bound=BaseModel)
type PostID = Annotated[str, AfterValidator(extract_post_id)]


class BaseEndpoint[T: BaseModel](ABC):
    OPERATOR: ClassVar[str]
    MODEL: type[T]

    FEATURES: ClassVar[dict[str, Any]] = {}
    VARIABLES: ClassVar[dict[str, Any]] = {}

    def __init__(self, api: TwitterAPI) -> None:
        self.api = api

    async def _list_request(
        self,
        data: dict[str, Any] | None = None,
    ) -> AsyncGenerator[list[dict[str, Any]]]:
        async for entries in self.api.list_graphql_request(
            self.OPERATOR,
            data={**self.VARIABLES, **(data or {})},
            features=self.FEATURES,
        ):
            yield entries

    async def _single_request(
        self,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any] | None:
        return await self.api.get_graphql_request(
            self.OPERATOR,
            data={**self.VARIABLES, **(data or {})},
            features=self.FEATURES,
        )

    def _serialize(self, data: Any) -> T:
        return self.MODEL.from_twitter_response(data)


class Users(BaseEndpoint[TwitterUser]):
    OPERATOR = '1VOOyvKkiI3FMmkeDNxM9A/UserByScreenName'

    MODEL = TwitterUser

    FEATURES = {
        'highlights_tweets_tab_ui_enabled': True,
        'hidden_profile_likes_enabled': True,
        'creator_subscriptions_tweet_preview_api_enabled': True,
        'hidden_profile_subscriptions_enabled': True,
        'subscriptions_verification_info_verified_since_enabled': True,
        'subscriptions_verification_info_is_identity_verified_enabled': False,
        'responsive_web_twitter_article_notes_tab_enabled': False,
        'subscriptions_feature_can_gift_premium': False,
        'profile_label_improvements_pcf_label_in_post_enabled': False,
    }

    async def get(self, username: str) -> TwitterUser | None:
        return self._serialize(
            await self._single_request(
                data={'screen_name': username, 'withSafetyModeUserFields': True},
            ),
        )


class Posts(BaseEndpoint[TwitterPost]):
    OPERATOR = '_8aYOgEDz35BrBcBal1-_w/TweetDetail'

    MODEL = TwitterPost

    VARIABLES = {
        'with_rux_injections': True,
        'includePromotedContent': True,
        'withCommunity': True,
        'withQuickPromoteEligibilityTweetFields': True,
        'withBirdwatchNotes': True,
        'withVoice': True,
        'withV2Timeline': True,
    }

    @validate_call
    async def get(self, post_id: PostID) -> TwitterPost | None:
        data = {'focalTweetId': str(post_id)}

        try:
            return self._serialize(await self._single_request(data=data))
        except ValidationError:
            logger.exception('Twitter API error')
            return None


class Likes(BaseEndpoint[TwitterPost]):
    OPERATOR = 'XHTMjDbiTGLQ9cP1em-aqQ/Likes'

    MODEL = TwitterPost

    VARIABLES = {
        'includePromotedContent': False,
        'withClientEventToken': False,
        'withBirdwatchNotes': False,
        'withVoice': True,
    }

    @validate_call
    async def list(
        self,
        user_id: UserID,
        count: int | None = None,
    ) -> AsyncGenerator[list[TwitterPost]]:
        data = {
            'userId': str(user_id),
            'count': count or 20,
        }

        async for entries in self._list_request(data=data):
            yield [self._serialize(entry) for entry in entries]

File: backend/src/services/twitter.ts
import UserAgent from "user-agents";
import { z } from "zod";

// Types
type UserID = string | number;
type PostID = string;

// Constants
const DEFAULT_KWARGS = {
	articles_preview_enabled: false,
	c9s_tweet_anatomy_moderator_badge_enabled: true,
	communities_web_enable_tweet_community_results_fetch: true,
	creator_subscriptions_quote_tweet_preview_enabled: false,
	creator_subscriptions_tweet_preview_api_enabled: true,
	freedom_of_speech_not_reach_fetch_enabled: true,
	graphql_is_translatable_rweb_tweet_is_translatable_enabled: true,
	longform_notetweets_consumption_enabled: true,
	longform_notetweets_inline_media_enabled: true,
	longform_notetweets_rich_text_read_enabled: true,
	responsive_web_edit_tweet_api_enabled: true,
	responsive_web_enhance_cards_enabled: false,
	responsive_web_graphql_exclude_directive_enabled: true,
	responsive_web_graphql_skip_user_profile_image_extensions_enabled: false,
	responsive_web_graphql_timeline_navigation_enabled: true,
	responsive_web_media_download_video_enabled: false,
	responsive_web_twitter_article_tweet_consumption_enabled: true,
	rweb_tipjar_consumption_enabled: true,
	rweb_video_timestamps_enabled: true,
	standardized_nudges_misinfo: true,
	tweet_awards_web_tipping_enabled: false,
	tweet_with_visibility_results_prefer_gql_limited_actions_policy_enabled: true,
	tweet_with_visibility_results_prefer_gql_media_interstitial_enabled: false,
	tweetypie_unmention_optimization_enabled: true,
	verified_phone_label_enabled: false,
	view_counts_everywhere_api_enabled: true,
	responsive_web_grok_analyze_button_fetch_trends_enabled: false,
	premium_content_api_read_enabled: false,
	profile_label_improvements_pcf_label_in_post_enabled: false,
	responsive_web_grok_share_attachment_enabled: false,
	responsive_web_grok_analyze_post_followups_enabled: false,
	responsive_web_grok_image_annotation_enabled: false,
	responsive_web_grok_analysis_button_from_backend: false,
	responsive_web_jetfuel_frame: false,
	rweb_video_screen_enabled: true,
	responsive_web_grok_show_grok_translated_post: true,
} as const;

// Schemas
const MediaTypeSchema = z.enum(["photo", "video", "animated_gif"]);

const PostMediaSchema = z
	.object({
		media_id: z.string(),
		media_key: z.string(),
		url: z.string(),
		media_type: MediaTypeSchema,
	})
	.transform((data) => ({
		mediaId: data.media_id,
		mediaKey: data.media_key,
		url: data.url,
		mediaType: data.media_type,
		get largeUrl() {
			return `${data.url}?format=jpg&name=large`;
		},
		get isPhoto() {
			return data.media_type === "photo";
		},
	}));

const TwitterUserSchema = z
	.object({
		rest_id: z.string(),
		legacy: z.object({
			name: z.string(),
			screen_name: z.string(),
			profile_image_url_https: z.string(),
		}),
		is_blue_verified: z.boolean(),
	})
	.transform((data) => ({
		userId: Number.parseInt(data.rest_id),
		name: data.legacy.name,
		username: data.legacy.screen_name,
		avatar: data.legacy.profile_image_url_https,
		isBlueVerified: data.is_blue_verified,
	}));

const TwitterPostSchema = z
	.object({
		rest_id: z.string().optional(),
		entryId: z.string().optional(),
		legacy: z
			.object({
				extended_entities: z
					.object({
						media: z.array(z.any()).optional(),
					})
					.optional(),
			})
			.optional(),
		core: z.object({
			user_results: z.object({
				result: z.any(),
			}),
		}),
	})
	.transform((data) => {
		const postId = (data.rest_id || data.entryId || "").replace(/^tweet-/, "");
		const media = data.legacy?.extended_entities?.media || [];
		const user = TwitterUserSchema.parse(data.core.user_results.result);

		return {
			postId,
			media: media.map((m: any) =>
				PostMediaSchema.parse({
					media_id: m.id_str,
					media_key: m.media_key,
					url: m.media_url_https,
					media_type: m.type,
				}),
			),
			user,
			get url() {
				return `https://x.com/${user.username}/status/${postId}`;
			},
		};
	});

export type TwitterUser = z.infer<typeof TwitterUserSchema>;
export type TwitterPost = z.infer<typeof TwitterPostSchema>;
export type PostMedia = z.infer<typeof PostMediaSchema>;

// Utility functions
function encodeParams(data: Record<string, any>): Record<string, string> {
	const result: Record<string, string> = {};

	for (const [key, value] of Object.entries(data)) {
		if (typeof value === "object" && value !== null) {
			const filteredValue = Object.fromEntries(
				Object.entries(value).filter(([, v]) => v !== null && v !== undefined),
			);
			result[key] = JSON.stringify(filteredValue);
		}
	}

	return result;
}

function findKeyRecursive(
	data: any,
	targetKey: string,
	targetPrefix?: string,
): any {
	const stack = [data];

	while (stack.length > 0) {
		const current = stack.pop();

		if (typeof current === "object" && current !== null) {
			if (Array.isArray(current)) {
				stack.push(...current);
			} else {
				if (targetKey in current) {
					const value = current[targetKey];
					if (targetPrefix === undefined) {
						return value;
					}

					if (typeof value === "string" && value.startsWith(targetPrefix)) {
						return current;
					}
				}

				stack.push(...Object.values(current));
			}
		}
	}

	return null;
}

function extractPostId(url: string): string {
	const match = url.match(/^https:\/\/x\.com\/(.*)\/status\/(\d+)(\?s=.*)?$/);

	if (!match) {
		throw new Error("Invalid post ID");
	}

	return match[2];
}

export class TwitterAPIError extends Error {
	constructor(
		public readonly statusCode: number,
		public readonly reason: string,
	) {
		super(`Unexpected error from Twitter API: ${statusCode} ${reason}`);
		this.name = "TwitterAPIError";
	}

	static fromResponse(response: Response, text: string): TwitterAPIError {
		return new TwitterAPIError(response.status, text);
	}
}

// Main API class
export class TwitterAPI {
	private readonly baseUrl: string;
	private readonly headers: Record<string, string>;

	constructor(
		cookies: Record<string, string>,
		baseUrl = "https://x.com/i/api/graphql/",
		userAgent?: string,
	) {
		this.baseUrl = baseUrl;
		this.headers = {
			"User-Agent": userAgent || new UserAgent().toString(),
			"x-twitter-client-language": "en",
			"x-twitter-active-user": "yes",
			Authorization:
				"Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA",
			"x-csrf-token": cookies.ct0,
			"x-twitter-auth-type": "OAuth2Session",
			Cookie: Object.entries(cookies)
				.map(([key, value]) => `${key}=${value}`)
				.join("; "),
		};
	}

	private static prepareParams(
		data?: Record<string, any>,
		features?: Record<string, any>,
	): Record<string, any> {
		const featuresPayload = { ...DEFAULT_KWARGS, ...features };

		return {
			variables: data || {},
			features: featuresPayload,
		};
	}

	async *listGraphqlRequest(
		url: string,
		data?: Record<string, any>,
		features?: Record<string, any>,
	): AsyncGenerator<any[]> {
		const params = TwitterAPI.prepareParams(data, features);
		let cursor: string | null = null;

		do {
			if (cursor) {
				params.variables.cursor = cursor;
			}

			const searchParams = new URLSearchParams(encodeParams(params));
			const fullUrl = `${this.baseUrl}${url}?${searchParams}`;

			const response = await fetch(fullUrl, {
				method: "GET",
				headers: this.headers,
			});

			if (!response.ok) {
				const text = await response.text();
				console.error("Twitter API error:", text);
				throw TwitterAPIError.fromResponse(response, text);
			}

			let jsonResponse: any;
			try {
				jsonResponse = await response.json();
			} catch (error) {
				const text = await response.text();
				throw TwitterAPIError.fromResponse(response, text);
			}

			console.debug("JSON response:", JSON.stringify(jsonResponse));

			const entries = findKeyRecursive(jsonResponse, "entries");

			if (!entries) {
				return;
			}

			const filteredEntries = entries.filter(
				(entry: any) =>
					entry?.entryId &&
					!entry.entryId.startsWith("cursor-") &&
					!entry.entryId.startsWith("messageprompt-"),
			);

			console.debug("Filtered entries:", filteredEntries);

			if (filteredEntries.length > 0) {
				yield filteredEntries;
			} else {
				return;
			}

			const cursorEntry = findKeyRecursive(
				jsonResponse,
				"entryId",
				"cursor-bottom",
			);

			if (!cursorEntry) {
				return;
			}

			cursor = cursorEntry.content?.itemContent?.value || null;
		} while (cursor);
	}

	async getGraphqlRequest(
		url: string,
		data?: Record<string, any>,
		features?: Record<string, any>,
	): Promise<any | null> {
		const params = TwitterAPI.prepareParams(data, features);
		const searchParams = new URLSearchParams(encodeParams(params));
		const fullUrl = `${this.baseUrl}${url}?${searchParams}`;

		const response = await fetch(fullUrl, {
			method: "GET",
			headers: this.headers,
		});

		if (response.status === 404) {
			return null;
		}

		if (!response.ok) {
			const text = await response.text();
			throw TwitterAPIError.fromResponse(response, text);
		}

		let jsonResponse: any;
		try {
			jsonResponse = await response.json();
		} catch (error) {
			const text = await response.text();
			throw TwitterAPIError.fromResponse(response, text);
		}

		console.debug("JSON response:", JSON.stringify(jsonResponse));

		return jsonResponse;
	}

	get users(): Users {
		return new Users(this);
	}

	get likes(): Likes {
		return new Likes(this);
	}

	get posts(): Posts {
		return new Posts(this);
	}
}

// Base endpoint class
abstract class BaseEndpoint<T> {
	protected abstract readonly operator: string;
	protected abstract readonly features: Record<string, any>;
	protected abstract readonly variables: Record<string, any>;

	constructor(protected readonly api: TwitterAPI) {}

	protected async *listRequest(
		data?: Record<string, any>,
	): AsyncGenerator<any[]> {
		const requestData = { ...this.variables, ...data };

		for await (const entries of this.api.listGraphqlRequest(
			this.operator,
			requestData,
			this.features,
		)) {
			yield entries;
		}
	}

	protected async singleRequest(
		data?: Record<string, any>,
	): Promise<any | null> {
		const requestData = { ...this.variables, ...data };

		return await this.api.getGraphqlRequest(
			this.operator,
			requestData,
			this.features,
		);
	}

	protected abstract serialize(data: any): T;
}

// Users endpoint
export class Users extends BaseEndpoint<TwitterUser> {
	protected readonly operator = "1VOOyvKkiI3FMmkeDNxM9A/UserByScreenName";
	protected readonly variables = {};
	protected readonly features = {
		highlights_tweets_tab_ui_enabled: true,
		hidden_profile_likes_enabled: true,
		creator_subscriptions_tweet_preview_api_enabled: true,
		hidden_profile_subscriptions_enabled: true,
		subscriptions_verification_info_verified_since_enabled: true,
		subscriptions_verification_info_is_identity_verified_enabled: false,
		responsive_web_twitter_article_notes_tab_enabled: false,
		subscriptions_feature_can_gift_premium: false,
		profile_label_improvements_pcf_label_in_post_enabled: false,
	};

	protected serialize(data: any): TwitterUser {
		const result = findKeyRecursive(data, "result");
		return TwitterUserSchema.parse(result);
	}

	async get(username: string): Promise<TwitterUser | null> {
		const data = await this.singleRequest({
			screen_name: username,
			withSafetyModeUserFields: true,
		});

		if (!data) {
			return null;
		}

		return this.serialize(data);
	}
}

// Posts endpoint
export class Posts extends BaseEndpoint<TwitterPost> {
	protected readonly operator = "_8aYOgEDz35BrBcBal1-_w/TweetDetail";
	protected readonly variables = {
		with_rux_injections: true,
		includePromotedContent: true,
		withCommunity: true,
		withQuickPromoteEligibilityTweetFields: true,
		withBirdwatchNotes: true,
		withVoice: true,
		withV2Timeline: true,
	};
	protected readonly features = {};

	protected serialize(data: any): TwitterPost {
		const firstPost = findKeyRecursive(data, "entries");

		if (!firstPost) {
			const result = findKeyRecursive(data, "result");
			return TwitterPostSchema.parse(result);
		}

		const result = findKeyRecursive(firstPost, "result");
		return TwitterPostSchema.parse(result);
	}

	async get(postIdOrUrl: string): Promise<TwitterPost | null> {
		let postId: string;

		try {
			postId = postIdOrUrl.includes("x.com")
				? extractPostId(postIdOrUrl)
				: postIdOrUrl;
		} catch (error) {
			console.error("Twitter API error:", error);
			return null;
		}

		const data = await this.singleRequest({
			focalTweetId: postId,
		});

		if (!data) {
			return null;
		}

		try {
			return this.serialize(data);
		} catch (error) {
			console.error("Twitter API error:", error);
			return null;
		}
	}
}

// Likes endpoint
export class Likes extends BaseEndpoint<TwitterPost> {
	protected readonly operator = "XHTMjDbiTGLQ9cP1em-aqQ/Likes";
	protected readonly variables = {
		includePromotedContent: false,
		withClientEventToken: false,
		withBirdwatchNotes: false,
		withVoice: true,
	};
	protected readonly features = {};

	protected serialize(data: any): TwitterPost {
		return TwitterPostSchema.parse(data);
	}

	async *list(userId: UserID, count?: number): AsyncGenerator<TwitterPost[]> {
		const data = {
			userId: String(userId),
			count: count || 20,
		};

		for await (const entries of this.listRequest(data)) {
			yield entries.map((entry) => this.serialize(entry));
		}
	}
}
